File: src/motion_types/scripts/straight_line.py
#!/usr/bin/env python3
# license removed for brevity
import rospy
from geometry_msgs.msg import Twist
from turtlesim.msg import Pose
import math
import time 



# Initialize global variables
x = 0
y = 0
yaw = 0

def move(velocity_publisher, speed, distance, is_forward):
    velocity_message = Twist()

    global x, y

    x0 = x
    y0 = y

    if is_forward:
        velocity_message.linear.x = abs(speed)
    else:
        velocity_message.linear.x = -abs(speed)
    
    distance_moved = 0.0
    loop_rate = rospy.Rate(100)

    while True:
        rospy.loginfo("Turtle is moving")
        velocity_publisher.publish(velocity_message)

        loop_rate.sleep()

        distance_moved = abs(math.sqrt(((x - x0)**2) + ((y - y0)**2)))
        print(distance_moved)

        if not (distance_moved < distance):
            rospy.loginfo("Reached the desired distance")
            break

    velocity_message.linear.x = 0
    velocity_publisher.publish(velocity_message)

def poseCallback(pose_message):
    global x, y, yaw
    x = pose_message.x
    y = pose_message.y
    yaw = pose_message.theta

if __name__ == '__main__':        
    rospy.init_node('Robot_move_straight', anonymous=True)

    cmd_vel_topic = '/turtle1/cmd_vel'
    velocity_publisher = rospy.Publisher(cmd_vel_topic, Twist, queue_size=10)

    position_topic = "/turtle1/pose"
    pose_subscriber = rospy.Subscriber(position_topic, Pose, poseCallback)
    time.sleep(2)

    move(velocity_publisher, 1.0, 5.0, False)

File: src/motion_types/src/straight_line.cpp
#include "ros/ros.h"
#include "geometry_msgs/Twist.h"
#include "turtlesim/Pose.h"
#include <iostream>
#include <cmath>

ros::Publisher velocity_publisher;
turtlesim::Pose turtlesim_pose;

void move(double speed, double distance, bool isForward) {
    geometry_msgs::Twist vel_msg;

    // Set linear velocity
    if (isForward)
        vel_msg.linear.x = abs(speed);
    else
        vel_msg.linear.x = -abs(speed);

    // No movement in y, z directions or angular velocity
    vel_msg.linear.y = 0;
    vel_msg.linear.z = 0;
    vel_msg.angular.x = 0;
    vel_msg.angular.y = 0;
    vel_msg.angular.z = 0;

    double t0 = ros::Time::now().toSec();
    double current_distance = 0.0;
    ros::Rate loop_rate(10);  // Control the loop speed

    do {
        velocity_publisher.publish(vel_msg);   // send x = speed 

        double t1 = ros::Time::now().toSec();

        current_distance = speed * (t1 - t0);  // Distance = velocity * time
        ros::spinOnce();
        loop_rate.sleep();
    } while (current_distance < distance);

    // Stop the turtle after reaching the desired distance
    vel_msg.linear.x = 0;
    velocity_publisher.publish(vel_msg);
}

void poseCallback(const turtlesim::Pose::ConstPtr &pose_message) {
    turtlesim_pose.x = pose_message->x;
    turtlesim_pose.y = pose_message->y;
    turtlesim_pose.theta = pose_message->theta;
}



int main(int argc, char **argv) {
    // Initiate new ROS node named "move_straight_line"
    ros::init(argc, argv, "move_straight_line");
    ros::NodeHandle n;

    //Initialize variables
    double speed, distance;
    bool isForward;

    // Define the publisher and subscriber
    velocity_publisher = n.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 1000);
    ros::Subscriber pose_subscriber = n.subscribe("/turtle1/pose", 10, poseCallback);

    // Test the code
    ROS_INFO("\n\n\n******START TESTING************\n");
    std::cout << "Enter speed: ";
    std::cin >> speed;
    std::cout << "Enter distance: ";
    std::cin >> distance;
    std::cout << "Forward (1 for yes, 0 for no)?: ";
    std::cin >> isForward;

    move(speed, distance, isForward);

    return 0;
}
